### ISSUE TITLE
The best way to read big SPSS (SAV) files?

### ISSUE THREAD

**Issue by gtamas** [Roche/pyreadstat#80]
This is just a question.

Reading a 260MB SPSS file takes about 60s using either the read_file_in_chunks or read_sav functions. I tried several different chunk sizes (1000, 10 000 etc) but there seem to be no big changes in performance.

So is this to be expected?  I mean, performance section in the readme promises 7s read speed (for a somewhat smaller file and a different format).

This is just simple test script:

```
import pyreadstat

reader = pyreadstat.read_file_in_chunks(pyreadstat.read_sav, './big.sav', chunksize= 1000)

for df, meta in reader:
   continue

print("ok")
```

So this takes about a minute using Python 3.9 on MacOSX Catalina, MacBook Pro. Is this normal? 

Is there any way to make reading the file faster? I'm not a Python guy, so maybe I made a mistake.

Thx and great library by the way!

**Comment by gtamas** [comment]
Sorry, the process takes longer. Just tested it, and it took 2 minutes to run.

**Comment by ofajardo** [comment]
The performance tests in the Readme are for SAS files, not for SPSS. I never profiled SPSS files.

I'll take a look to see if there is any opportunity to improve the performance, but honestly speaking I think it's a bit unlikely to improve much. 

How many rows and columns does your dataset have and what data types do the columns have? Even better if you could pass the code to generate a test file using pyreadstat.write_spss.

**Comment by ofajardo** [comment]
just tested a 160 Mb Sav file, with 200 K rows and 100 columns and it was read in 12 seconds in  my old machine. That looks quite reasonable for me. Therefore you must provide a sample file that is causing you issues so that I can investigate further. As I guess you cannot share your original data, you can make a synthetic dataset using pyreadstat.write_sav. Look at your original data and think on what can be causing the problems. Things to look at are the numbers of columns you have (for SAS files I remember somebody mentioning when you have a large amount of columns it becomes slow), or data types (dates for example are slower to read). 

Here as an example how I created the 200K rows x 100 cols dataset:

```
>>> import pyreadstat
>>> import numpy as np
>>> r = np.random.rand(200000, 100)
>>> cols = ["col"+str(x) for x in range(0,100)]
>>> df = pd.DataFrame(r, columns=cols)
>>> pyreadstat.write_sav(df, "big.sav")

```

**Comment by ofajardo** [comment]
just tested a dataset of 100K rows x 1000 columns and it took again 12 seconds. 

Please provide a sample file.

**Comment by gtamas** [comment]
OK Thanks you for looking into.this!
I will obtain a test a file next week.

By the the way: the file we are talking about has 294 000 rows and 666 columns, and it's about 225MB.

I will get back to regarding the sample file asap.

**Comment by gtamas** [comment]
@ofajardo 
Just tried your random file maker code, but there is an undefined varibale:
```

Traceback (most recent call last):
  File "/some/path/test_write.py", line 5, in <module>
    df = pd.DataFrame(r, columns=cols)
NameError: name 'pd' is not defined
```

Is there a typo? Where the pd variable is supposed to be defined?

**Comment by ofajardo** [comment]
pd is the convention for pandas: 

```
import pandas as pd
```

**Comment by ofajardo** [comment]
Btw if you run that code, and then read the resulting file, how long does it take for you?

**Comment by gtamas** [comment]
OK I just tested it.  First I ran your code, which created a 153M file. 

Then I tried to read that file and it only took 6s.

Next, I ran the code below to generate a file similar to ours.

```
import pyreadstat
import numpy as np
import pandas as pd
r = np.random.rand(294000, 666)
cols = ["col"+str(x) for x in range(0,666)]
df = pd.DataFrame(r, columns=cols)
pyreadstat.write_sav(df, "random.sav")
```

This resulted in a 1.5GB file which I tried to read using:

```
import pyreadstat

reader = pyreadstat.read_file_in_chunks(pyreadstat.read_sav, './random.sav', chunksize= 10000)

for df, meta in reader:
   continue

print("ok")
```

The process takes abut 60s. I monitored the time using top (unix command).

Don't get it. So it seems this random file can be parsed very fast, but our file takes way longer. Our file was created by SPSS  and there is nothing special in it, just a lot of data. I'm confused. Is this normal or should parsing be faster?

I will try to get permission to give the file to you, so you can test. I guess that would be the best.

**Comment by ofajardo** [comment]
So you said your file of 250 Mb and 300K rows x 666 columns is read in 60s, and the synthetic file of 300K rows x 666 columns and 1.5 GB is also read in 60 s. Then it makes sense, the sizes of both datasets is the same and the time to read them is the same. If you are wondering why your file is 250 Mb and the synthetic 1.5 Gb ... Maybe yours is compressed. You can try to compress the synthetic as well:

```
pyreadstat.write_sav(df, "big.sav", compress=True)
```

**Comment by ofajardo** [comment]
Apparently Spss saves files as compressed by default: 
https://www.ibm.com/support/knowledgecenter/SSLVMB_24.0.0/spss/base/syn_save_compressed_uncompressed.html 
That would explain the difference in size of the files. Pyreadstat does Z compression that is even more compact.

**Comment by ofajardo** [comment]
Maybe the way to improve this is to parallelize the task. In my mac reading the 294000x666 file takes 115 seconds if I do it in one go. By parallelizing it I get 29 seconds: 

first I create a file worker.py:

```
def worker(inpt):
    import pyreadstat
    offset, chunksize, path = inpt
    df, meta = pyreadstat.read_sav(path, row_offset=offset, row_limit=chunksize)
    return df
```

and then in the main script I have this:

```
import multiprocessing as mp
from time import time

import pandas as pd
import pyreadstat

from worker import worker

# calculate the number of rows in the file
_, meta = pyreadstat.read_sav("big.sav", metadataonly=True)
numrows = meta.number_rows
# calculate number of cores in the machine, this could also be set manually to some number, i.e. 8
numcores = mp.cpu_count()
# calculate the chunksize and offsets
divs = [numrows // numcores + (1 if x < numrows % numcores else 0)  for x in range (numcores) ]
chunksize = divs[0]
offsets = [indx*chunksize for indx in range(numcores)] 
# pack the data for the jobs
jobs = [(x, chunksize, "big.sav") for x in offsets]

pool = mp.Pool(processes=numcores)
# let's go!
t0=time()
chunks = pool.map(worker, jobs)
t1=time()
print(t1-t0) # this prints 29 seconds
# you can concatenate all the chunks into a single big dataframe if you like
#final = pd.concat(chunks, axis=0, ignore_index=True)
```

How does this work for you? If it improves things I can investigate how to refine this and pack it into a user friendlier function as I did with read_file_in_chunks.

**Comment by ofajardo** [comment]
I also found another bottleneck in the base code that I have been able to improve. Now for the big file, when reading it in one go I improved from 157 to 93 seconds and when reading in multiple processes from 32 to 19.5 secs. If you take the extremes it will be an improvement from 157 to 19.5 seconds.

**Comment by ofajardo** [comment]
The new function read_file_multiprocessing is already present in the dev_performance1 branch in case you would like to test it.

**Comment by gtamas** [comment]
Thanks so much for the help! I will test these with some other files too. 
These are super improvements indeed! I this parallelisation and the improvements you just made are far enough for us. 
I will test these with some other files and will let you know how it worked. 

Btw, the bottleneck you mentioned is only fixed in dev_performance1 branch for now, right? So it's not part of the stable code yet.

Thanks again!

**Comment by ofajardo** [comment]
Right, all changes are in the dev_performance1 branch. 
It would be very nice to have some feedback before releasing, thanks!

**Comment by ofajardo** [comment]
good news! I managed to get another big improvement. Now the big file is read in 32 seconds if in one go and in 11 seconds if using parallelism. Changes are still in dev_performance1 branch for the moment.

**Comment by ofajardo** [comment]
ok managed to get it down to 28 secs in one go and 10 secs using parallelism. I think that's as good as you can get it for now, so I released the code in the new version 1.0.3.

I'm going to close this, but it would be awesome if you provide some feedback!
